wrappers/rust/icicle-core/src/polynomials/mod.rs

```diff
@@ -0,0 +1,707 @@
+        impl Mul<&UnivariatePolynomial> for &$field {
+            type Output = UnivariatePolynomial;
+
+            fn mul(self, rhs: &UnivariatePolynomial) -> Self::Output {
+                unsafe {
+                    UnivariatePolynomial {
+                        handle: multiply_by_scalar(rhs.handle, self),
+                    }
+                }
+            }
+        }
+
+        impl Div for &UnivariatePolynomial {
+            type Output = UnivariatePolynomial;
+
+            fn div(self: Self, rhs: Self) -> Self::Output {
+                unsafe {
+                    UnivariatePolynomial {
+                        handle: quotient(self.handle, rhs.handle),
+                    }
+                }
+            }
+        }
+
+        impl Rem for &UnivariatePolynomial {
+            type Output = UnivariatePolynomial;
+
+            fn rem(self: Self, rhs: Self) -> Self::Output {
+                unsafe {
+                    UnivariatePolynomial {
+                        handle: remainder(self.handle, rhs.handle),
+                    }
+                }
+            }
+        }
+    };
+}
+
+#[macro_export]
+macro_rules! impl_polynomial_tests {
+    (
+        $field_prefix_ident:ident,
+        $field:ident
+    ) => {
+        use super::*;
+        use icicle_core::ntt::{get_root_of_unity, initialize_domain, release_domain, NTTDomain};
+        use icicle_core::vec_ops::{add_scalars, mul_scalars, sub_scalars, VecOps, VecOpsConfig};
+        use icicle_cuda_runtime::device_context::DeviceContext;
+        use icicle_cuda_runtime::memory::{DeviceVec, HostSlice};
+        use std::sync::Once;
+
+        use icicle_core::traits::{FieldImpl, GenerateRandom};
+
+        type Poly = UnivariatePolynomial;
+
+        fn init_domain<F: FieldImpl>(max_size: u64, device_id: usize, fast_twiddles_mode: bool)
+        where
+            <F as FieldImpl>::Config: NTTDomain<F>,
+        {
+            let ctx = DeviceContext::default_for_device(device_id);
+            let rou: F = get_root_of_unity(max_size);
+            initialize_domain(rou, &ctx, fast_twiddles_mode).unwrap();
+        }
+
+        fn rel_domain<F: FieldImpl>(device_id: usize)
+        where
+            <F as FieldImpl>::Config: NTTDomain<F>,
+        {
+            let ctx = DeviceContext::default_for_device(device_id);
+            release_domain::<F>(&ctx).unwrap()
+        }
+
+        fn randomize_coeffs<F: FieldImpl>(size: usize) -> Vec<F>
+        where
+            <F as FieldImpl>::Config: GenerateRandom<F>,
+        {
+            let coeffs = F::Config::generate_random(size);
+            coeffs
+        }
+
+        fn rand() -> $field {
+            let r = randomize_coeffs::<$field>(1);
+            // let coeffs = $field::Config::generate_random(1);
+            r[0]
+        }
+
+        // Note: implementing field arithmetic (+,-,*) for fields via vec_ops since they are not implemented on host
+        fn add(a: &$field, b: &$field) -> $field {
```

Comment:
Can we move it into tests.rs as a templated function? (the same goes for all the other tests). Motivation - it's easier to work on tests that are not inside macro, for this reason we should keep as little code inside macros as possible

Reply:
I can try again yes

Reply:
see my answer above, to Jeremy's comment